npm/nwpolicy.go

```diff
@@ -13,62 +14,118 @@
 	networkingv1 "k8s.io/api/networking/v1"
 )
 
+// GetNetworkPolicyKey will return netpolKey
+func GetNetworkPolicyKey(npObj *networkingv1.NetworkPolicy) string {
+	netpolKey, err := util.GetObjKeyFunc(npObj)
+	if err != nil {
+		metrics.SendErrorLogAndMetric(util.NetpolID, "[GetNetworkPolicyKey] Error: while running MetaNamespaceKeyFunc err: %s", err)
+		return ""
+	}
+	if len(netpolKey) == 0 {
+		return ""
+	}
+	return util.GetNSNameWithPrefix(netpolKey)
+}
+
+// GetProcessedNPKey will return netpolKey
+func GetProcessedNPKey(npObj *networkingv1.NetworkPolicy, hashSelector string) string {
+	// hashSelector will never be empty
+	netpolKey := hashSelector
+	if len(npObj.GetNamespace()) > 0 {
+		netpolKey = npObj.GetNamespace() + "/" + netpolKey
+	}
+	return util.GetNSNameWithPrefix(netpolKey)
+}
+
 func (npMgr *NetworkPolicyManager) canCleanUpNpmChains() bool {
 	if !npMgr.isSafeToCleanUpAzureNpmChain {
 		return false
 	}
 
-	for _, ns := range npMgr.NsMap {
-		if len(ns.ProcessedNpMap) > 0 {
-			return false
-		}
+	if len(npMgr.ProcessedNpMap) > 0 {
+		return false
 	}
 
 	return true
 }
 
+func (npMgr *NetworkPolicyManager) policyExists(npObj *networkingv1.NetworkPolicy) bool {
+	npKey := GetNetworkPolicyKey(npObj)
+	if npKey == "" {
+		return false
+	}
+
+	np, exists := npMgr.RawNpMap[npKey]
+	if !exists {
+		return false
+	}
+
+	if !util.CompareResourceVersions(np.ObjectMeta.ResourceVersion, npObj.ObjectMeta.ResourceVersion) {
+		log.Logf("Cached Network Policy has larger ResourceVersion number than new Obj. Name: %s Cached RV: %d New RV: %d\n",
+			npObj.ObjectMeta.Name,
+			np.ObjectMeta.ResourceVersion,
+			npObj.ObjectMeta.ResourceVersion,
+		)
+		return true
+	}
+
+	if isSamePolicy(np, npObj) {
+		return true
+	}
+
+	return false
+}
+
 // AddNetworkPolicy handles adding network policy to iptables.
 func (npMgr *NetworkPolicyManager) AddNetworkPolicy(npObj *networkingv1.NetworkPolicy) error {
 	var (
-		err    error
-		ns     *Namespace
-		exists bool
-		npNs   = util.GetNSNameWithPrefix(npObj.ObjectMeta.Namespace)
-		npName = npObj.ObjectMeta.Name
-		allNs  = npMgr.NsMap[util.KubeAllNamespacesFlag]
-		timer  = metrics.StartNewTimer()
+		err            error
+		ns             *Namespace
+		exists         bool
+		npNs           = util.GetNSNameWithPrefix(npObj.ObjectMeta.Namespace)
+		npName         = npObj.ObjectMeta.Name
+		allNs          = npMgr.NsMap[util.KubeAllNamespacesFlag]
+		timer          = metrics.StartNewTimer()
+		hashedSelector = HashSelector(&npObj.Spec.PodSelector)
+		npKey          = GetNetworkPolicyKey(npObj)
+		npProcessedKey = GetProcessedNPKey(npObj, hashedSelector)
```

Comment:
What is the difference between processed NP key and raw NP policy, why cant they be same?

Reply:
processed Key is based on the pod selector, if different netpols with different names have same pod selector, we are creating a single processedNp aggregating all the rules for the podselector on which we are applying